components/script/dom/notification.rs

```diff
@@ -0,0 +1,444 @@
+
+/// <https://notifications.spec.whatwg.org/#notifications>
+#[dom_struct]
+pub(crate) struct Notification {
+    eventtarget: EventTarget,
+    /// <https://notifications.spec.whatwg.org/#service-worker-registration>
+    serviceworker_registration: Option<Dom<ServiceWorkerRegistration>>,
+    /// <https://notifications.spec.whatwg.org/#concept-title>
+    title: DOMString,
+    /// <https://notifications.spec.whatwg.org/#body>
+    body: DOMString,
+    /// <https://notifications.spec.whatwg.org/#data>
+    #[ignore_malloc_size_of = "mozjs"]
+    data: Box<Heap<JSVal>>,
+    /// <https://notifications.spec.whatwg.org/#concept-direction>
+    dir: NotificationDirection,
+    /// <https://notifications.spec.whatwg.org/#image-url>
+    image: Option<USVString>,
+    /// <https://notifications.spec.whatwg.org/#icon-url>
+    icon: Option<USVString>,
+    /// <https://notifications.spec.whatwg.org/#badge-url>
+    badge: Option<USVString>,
+    /// <https://notifications.spec.whatwg.org/#concept-language>
+    lang: DOMString,
+    /// <https://notifications.spec.whatwg.org/#silent-preference-flag>
+    silent: Option<bool>,
+    /// <https://notifications.spec.whatwg.org/#tag>
+    tag: DOMString,
+    /// <https://notifications.spec.whatwg.org/#concept-origin>
+    #[no_trace] // ImmutableOrigin is not traceable
+    origin: ImmutableOrigin,
+    /// <https://notifications.spec.whatwg.org/#vibration-pattern>
+    vibration_pattern: Vec<u32>,
+    /// <https://notifications.spec.whatwg.org/#timestamp>
+    timestamp: u64,
+    /// <https://notifications.spec.whatwg.org/#renotify-preference-flag>
+    renotify: bool,
+    /// <https://notifications.spec.whatwg.org/#require-interaction-preference-flag>
+    require_interaction: bool,
+    /// <https://notifications.spec.whatwg.org/#actions>
+    #[ignore_malloc_size_of = "NotificationAction"]
+    actions: Vec<NotificationAction>,
+    // TODO: image resource, icon resource, and badge resource
+}
+
+impl Notification {
+    #[cfg_attr(crown, allow(crown::unrooted_must_root))]
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(
+        global: &GlobalScope,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+        origin: Option<ImmutableOrigin>,
+        base_url: Option<ServoUrl>,
+        fallback_timestamp: u64,
+    ) -> Fallible<DomRoot<Self>> {
+        let notification = Notification::new_inherited(
+            global,
+            title,
+            options,
+            origin,
+            base_url,
+            fallback_timestamp,
+        )?;
+        Ok(reflect_dom_object_with_proto(
+            Box::new(notification),
+            global,
+            proto,
+            can_gc,
+        ))
+    }
+
+    /// <https://notifications.spec.whatwg.org/#create-a-notification-with-a-settings-object>
+    fn new_inherited(
```

Comment:
This seems to correspond to `https://notifications.spec.whatwg.org/#create-a-notification`, and you are running one step of `https://notifications.spec.whatwg.org/#create-a-notification-with-a-settings-object` in the constructor. Would be good to restructure to match the spec, in ways that make sense and still using our `new` and `new_inherited` conventions. Also, you could not the TODO using the environment object, which is not implemented as such in servo.